File: 1501~2000/1835. Find XOR Sum of All Pairs Bitwise AND/main.py
"""
Source: https://rb.gy/0n3t3
Date: 2023/8/18
Skill:
Ref:
Runtime: 2552 ms, faster than 13.08%
Memory Usage: 30.96 MB, less than 100.00%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def getXORSum(self, arr1: List[int], arr2: List[int]) -> int:
        ones, res = [0 for _ in range(32)], 0
        for num in arr1:
            for i in range(32):
                if (num >> i) & 1 == 1:
                    ones[i] += 1
        for num in arr2:
            n = 0
            for i in range(32):
                if (num >> i) & 1 == 1 and ones[i] % 2 == 1:
                    n += 1 << i
            res ^= n
        return res


if __name__ == "__main__":
    s = Solution()

File: 1501~2000/1835. Find XOR Sum of All Pairs Bitwise AND/main.cpp
/**
 * Source: https://rb.gy/0n3t3
 * Date: 2024/4/5
 * Skill:
 * Ref:
 * Runtime: 94 ms, faster than 79.12% of C++ online submissions
 * Memory Usage: 95.76 MB, less than 92.77% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(1)
 * Constraints:
 *
 */


#include <iostream>
#include <vector>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;
#define SZ(X) ((int)(X).size())
#define ll long long
typedef unsigned long long ULL;

const int MX = 26;


class Solution {
public:
    int getXORSum(vector<int>& arr1, vector<int>& arr2) {
        int num = 0, res = 0;
        for (auto &i: arr2) num ^= i;
        for (auto &i: arr1) res ^= i & num;
        return res;
    }
};